src/crypto/x25519/mod.rs

```diff
@@ -549,19 +549,38 @@ fn mod_25519(x: Felem2) -> Felem {
 }
 
 fn mod_final_25519(x: Felem) -> Felem {
+    const MASK_64BITS: u128 = 0xffff_ffff_ffff_ffff;
+
     let mut acc0 = u128::from(x.0[0]);
     let mut acc1 = u128::from(x.0[1]);
     let mut acc2 = u128::from(x.0[2]);
     let mut acc3 = u128::from(x.0[3]);
 
     let mut top = acc3 >> 63;
     acc3 &= 0x7fff_ffff_ffff_ffff;
+    top = top.wrapping_mul(19);
+    acc0 = acc0.wrapping_add(top);
+    acc1 = acc1.wrapping_add(acc0 >> 64);
+    acc2 = acc2.wrapping_add(acc1 >> 64);
+    acc3 = acc3.wrapping_add(acc2 >> 64);
+
+    // Mask
+    acc0 &= MASK_64BITS;
+    acc1 &= MASK_64BITS;
+    acc2 &= MASK_64BITS;
+    acc3 &= MASK_64BITS;
 
+    top = acc0.wrapping_add(19) >> 64;
+    top = acc1.wrapping_add(top) >> 64;
+    top = acc2.wrapping_add(top) >> 64;
+    top = acc3.wrapping_add(top) >> 63;
     top = top.wrapping_mul(19);
+
     acc0 = acc0.wrapping_add(top);
     acc1 = acc1.wrapping_add(acc0 >> 64);
     acc2 = acc2.wrapping_add(acc1 >> 64);
     acc3 = acc3.wrapping_add(acc2 >> 64);
+    acc3 &= 0x7fff_ffff_ffff_ffff;
```

Comment:
What is this extra code doing? It doesn't make sense to me

Reply:
Looks like masking the top bit of the output. However is it not sufficient to execute just a single extra round of mul + add?

Reply:
I put the comments in the source code. What do you think?